pajbot/managers/irc.py

```diff
@@ -1,63 +1,155 @@
         self.bot = bot
 
-        self.channels = []
+        self.conn = None
+        self.ping_task = None
 
-        self.channel = "#" + self.bot.streamer
-        self.channels.append(self.channel)
+        self.num_privmsg_sent = 0
+        self.num_whispers_sent_minute = 0
+        self.num_whispers_sent_second = 0
 
-        self.control_hub_channel = self.bot.config["main"].get("control_hub", None)
-        if self.control_hub_channel:
-            self.control_hub_channel = "#" + self.control_hub_channel
-            self.channels.append(self.control_hub_channel)
+        streamer_channel = "#" + self.bot.streamer
+        self.channels = [streamer_channel]
 
-        chub = bot.config["main"].get("control_hub", "")
+        control_hub_channel = self.bot.config["main"].get("control_hub", None)
+        if control_hub_channel is not None:
+            self.channels.append("#" + control_hub_channel)
 
-        self.connection_manager = ConnectionManager(
-            self.bot.reactor,
-            self.bot,
-            streamer=self.bot.streamer,
-            control_hub_channel=chub,
-            host="irc.chat.twitch.tv",
-            port=6697,
-        )
+        bot.reactor.add_global_handler("all_events", self._dispatcher, -10)
 
+    @RateLimiter(max_calls=1, period=2)
     def start(self):
-        self.connection_manager.start()
-
-    def whisper(self, username, message, increase_message=True):
-        self.connection_manager.privmsg(
-            f"#{self.bot.nickname}", f"/w {username} {message}", increase_message=increase_message, whisper=True
-        )
-
-    def privmsg(self, message, channel, increase_message=True):
-        self.connection_manager.privmsg(channel, message, increase_message=increase_message)
-
-    def on_disconnect(self, chatconn, event):
-        self.connection_manager.on_disconnect(chatconn)
-
-    def _dispatcher(self, connection, event):
-        method = getattr(self.bot, "on_" + event.type, do_nothing)
         try:
-            method(connection, event)
+            self.make_new_connection()
         except:
-            log.exception("Logging an uncaught exception (IRC event handler)")
-
-    def on_welcome(self, conn, event):
+            log.exception("Failed to open connection, retrying in 2 seconds")
+            self.bot.execute_delayed(2, lambda: self.start())
+
+    def make_new_connection(self):
+        self.conn = Connection(self.bot.reactor)
+        with self.bot.reactor.mutex:
+            self.bot.reactor.connections.append(self.conn)
+        self.conn.connect(
+            "irc.chat.twitch.tv",
+            6697,
+            self.bot.nickname,
+            self.bot.password,
+            self.bot.nickname,
+            connect_factory=Factory(wrapper=ssl.wrap_socket),
+        )
+        self.conn.cap("REQ", "twitch.tv/commands", "twitch.tv/tags")
+
+        self.ping_task = ScheduleManager.execute_every(30, lambda: self.bot.execute_now(self._send_ping))
```

Comment:
This uses ScheduleManager, because you can't cancel tasks submitted via `bot.execute_every`.